ext/vulnsrc/suse/suse.go

```diff
@@ -0,0 +1,473 @@
+					Name: name(definition),
+					Link: link(definition),
+					//TODO: handle that once openSUSE/SLE OVAL files have severity info
+					Severity:    database.UnknownSeverity,
+					Description: description(definition),
+				},
+			}
+			for _, p := range pkgs {
+				vulnerability.Affected = append(vulnerability.Affected, p)
+			}
+			vulnerabilities = append(vulnerabilities, vulnerability)
+		}
+	}
+
+	return
+}
+
+func getCriterions(node criteria) [][]criterion {
+	// Filter useless criterions.
+	var criterions []criterion
+	for _, c := range node.Criterions {
+		ignored := false
+
+		for _, ignoredItem := range ignoredCriterions {
+			if strings.Contains(c.Comment, ignoredItem) {
+				ignored = true
+				break
+			}
+		}
+
+		if !ignored {
+			criterions = append(criterions, c)
+		}
+	}
+
+	if node.Operator == "AND" {
+		return [][]criterion{criterions}
+	} else if node.Operator == "OR" {
+		var possibilities [][]criterion
+		for _, c := range criterions {
+			possibilities = append(possibilities, []criterion{c})
+		}
+		return possibilities
+	}
+
+	return [][]criterion{}
+}
+
+func getPossibilities(node criteria) [][]criterion {
+	if len(node.Criterias) == 0 {
+		return getCriterions(node)
+	}
+
+	var possibilitiesToCompose [][][]criterion
+	for _, criteria := range node.Criterias {
+		possibilitiesToCompose = append(possibilitiesToCompose, getPossibilities(*criteria))
+	}
+	if len(node.Criterions) > 0 {
+		possibilitiesToCompose = append(possibilitiesToCompose, getCriterions(node))
+	}
+
+	var possibilities [][]criterion
+	if node.Operator == "AND" {
+		for _, possibility := range possibilitiesToCompose[0] {
+			possibilities = append(possibilities, possibility)
+		}
+
+		for _, possibilityGroup := range possibilitiesToCompose[1:] {
+			var newPossibilities [][]criterion
+
+			for _, possibility := range possibilities {
+				for _, possibilityInGroup := range possibilityGroup {
+					var p []criterion
+					p = append(p, possibility...)
+					p = append(p, possibilityInGroup...)
+					newPossibilities = append(newPossibilities, p)
+				}
+			}
+
+			possibilities = newPossibilities
+		}
+	} else if node.Operator == "OR" {
+		for _, possibilityGroup := range possibilitiesToCompose {
+			for _, possibility := range possibilityGroup {
+				possibilities = append(possibilities, possibility)
+			}
+		}
+	}
+
+	return possibilities
+}
+
+func toFeatureVersions(criteria criteria, osFlavor, osVersion string) []database.AffectedFeature {
+	// There are duplicates in SUSE .xml files.
+	// This map is for deduplication.
+	featureVersionParameters := make(map[string]database.AffectedFeature)
+
+	possibilities := getPossibilities(criteria)
+	for _, criterions := range possibilities {
+		var featureVersion database.AffectedFeature
+
+		// Attempt to parse package data from trees of criterions.
+		for _, c := range criterions {
+			if match := suseInstalledCommentRegexp.FindStringSubmatch(c.Comment); match != nil {
+				if len(match) != 4 {
```

Comment:
it's easier to read this conditional if you keep it `if (positive case) else negative case`.

```go
if len(match) == 4 {
  osVerison = ...
} else {
  log.WithField().Warning()
}
```